FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1494 @@
+                       Command_t * pCommand );
+
+/**
+ * @brief Remove an operation from the list of pending acks and return it.
+ *
+ * @param[in] usPacketId Packet ID of incoming ack.
+ * @param[in] xRemove Flag indicating if an ack should be removed
+ *
+ * @return Stored information about the operation awaiting the ack.
+ */
+static AckInfo_t prvGetAck( uint16_t usPacketId,
+                            bool xRemove );
+
+/**
+ * @brief Add a subscription to the subscription list.
+ *
+ * @note Multiple tasks can be subscribed to the same topic. However, a single
+ * task may only subscribe to the same topic filter once.
+ *
+ * @param[in] pTopicFilter Topic filter of subscription.
+ * @param[in] topicFilterLength Length of topic filter.
+ * @param[in] pxQueue Response queue in which to enqueue received publishes.
+ */
+static void prvAddSubscription( const char * pTopicFilter,
+                                uint16_t topicFilterLength,
+                                QueueHandle_t pxQueue );
+
+/**
+ * @brief Remove a subscription from the subscription list.
+ *
+ * @note If the topic filter exists multiple times in the subscription list,
+ * then every instance of the subscription will be removed.
+ *
+ * @param[in] pTopicFilter Topic filter of subscription.
+ * @param[in] topicFilterLength Length of topic filter.
+ */
+static void prvRemoveSubscription( const char * pTopicFilter,
+                                   uint16_t topicFilterLength );
+
+/**
+ * @brief Populate the parameters of a #Command_t
+ *
+ * @param[in] xCommandType Type of command.
+ * @param[in] pxContext Context and necessary structs for command.
+ * @param[in] xCallback Callback for when command completes.
+ * @param[out] pxCommand Pointer to initialized command.
+ *
+ * @return `true` if all necessary structs for the command exist in pxContext,
+ * else `false`
+ */
+static bool prvCreateCommand( CommandType_t xCommandType,
+                              CommandContext_t * pxContext,
+                              CommandCallback_t xCallback,
+                              Command_t * pxCommand );
+
+/**
+ * @brief Add a command to the global command queue.
+ *
+ * @param[in] pxCommand Pointer to command to copy to queue.
+ */
+static void prvAddCommandToQueue( Command_t * pxCommand );
+
+/**
+ * @brief Copy an incoming publish to a response queue.
+ *
+ * @param[in] pPublishInfo Info of incoming publish.
+ * @param[in] pResponseQueue Queue to which the publish is copied.
+ */
+static void prvCopyPublishToQueue( MQTTPublishInfo_t * pPublishInfo,
+                                   QueueHandle_t pResponseQueue );
+
+/**
+ * @brief Process a #Command_t.
+ *
+ * @note This demo does not check existing subscriptions before sending a
+ * SUBSCRIBE or UNSUBSCRIBE packet. If a subscription already exists, then
+ * a SUBSCRIBE packet will be sent anyway, and if multiple tasks are subscribed
+ * to a topic filter, then they will all be unsubscribed after an UNSUBSCRIBE.
```

Comment:
Is there a reason why we did it this way? can't we optimize the network traffic, by not sending a SUBSCRIBE, if a subscription already exists?

Reply:
A SUBSCRIBE packet shouldn't be huge, and doing so would be overly complicated for this demo and detract from what we're actually trying to demo. For example, If Task1 is subscribed to Topic2, and Task2 wants to subscribe to Topic1, Topic2, and Topic3 in the same SUBSCRIBE packet, we'd need to rewrite the buffer to only subscribe to Topic1 and Topic3. Also, sending a SUBSCRIBE for a topic you're already subscribed to is a valid use case (e.g. if you wanted to downgrade the maximum QoS).

Reply:
Ok Sounds good to me.